sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -443,31 +443,63 @@ object DataSourceStrategy {
     case expressions.EqualTo(Literal(v, t), a: Attribute) =>
       Some(sources.EqualTo(a.name, convertToScala(v, t)))
 
+    case expressions.EqualTo(Cast(c, dt, tz), Literal(v, t)) if c.isInstanceOf[Attribute] =>
+      Some(sources.EqualTo(c.asInstanceOf[Attribute].name, convertToScala(v, t)))
+    case expressions.EqualTo(Literal(v, t), Cast(c, dt, tz)) if c.isInstanceOf[Attribute] =>
+      Some(sources.EqualTo(c.asInstanceOf[Attribute].name, convertToScala(v, t)))
```

Comment:
I don't think this is correct. It can potentially end up with a filter with having a different type of the value in the pushed filter.

If you want to handle the cast, it will end up with the same fix as https://github.com/apache/spark/pull/8718 which was virtually rejected.

Reply:
cc @liancheng FYI